### DISCUSSION TITLE
Hinting the compiler at decreasing return values

### DISCUSSION THREAD

**Discussion by chtenb** [koka-lang/koka#450] [Q&A]
Consider the following code

```koka
abstract value struct vslice<a>( vec : vector<a>, start : int, len : int )

pub fun vslice/at( slice : vslice<a>, i : int ) : maybe<a>
  slice.vec.at(slice.start + i)

pub fun vslice/next( slice : vslice<a> ) : maybe<(a,vslice<a>)>
  match slice.at(0)
    Nothing -> Nothing
    Just(a) -> Just((a, slice(start = slice.start + 1, len = slice.len - 1)))
  
fun vslice/foreach( slice : vslice<a>, action : a -> e () ) : e ()
  match slice.next
    Nothing -> ()
    Just((x, xs)) -> 
      action(x)
      foreach(xs, action)
```

This fails to compile (unsurprisingly) because the compiler can't prove that the foreach function converges.
```
repro.kk(16, 5): error: effects do not match
 context        :     vslice/foreach
 term           :     vslice/foreach
 inferred effect: <div|_e1>
 expected effect: $e
```

This can be fixed by hinting the compiler that the recursive call is decreasing:
```
      foreach(pretend-decreasing(xs), action)
```

However, such an intervention is needed in every function that recurses using the `next` function. 
*Question*: Is it possible to instead place a hint in the `next` definition? This would make sense because there you already know that the result of `next` is decreasing. I tried doing this naively by 

```koka
    Just(a) -> Just((a, pretend-decreasing(slice(start = slice.start + 1, len = slice.len - 1))))
```

but that doesn't work.

**Comment by TimWhiting** [comment]
I would assume that you would want to use `map` or `foreach` or other functions that hide this as much as possible.

In other words this encourages users to avoid using the low level api `next` or just accept the div effect.

If you find that you often need to use the low level api, I would consider adding effects to get what you need. (For example if you want to break out `foreach` with a result I would consider adding a `break-result` effect or something as part of your vector library). Then when you use it you can just do a `foreach`, and use effects to jump out of the context or keep some state rather than using the low level `next` api. You could even make `next` an internal implementation detail, and explicitly say to use `foreach` for `for` loop like behavior.

In general, I think we are aware that the divergence detection could use more work to improve it. However this is less of a priority at this time compared to other things like adding back async support, improving the standard library, and improving the compilation speed / memory issues.

I understand that my view might not always be the desired solution, but in general proving non-divergence is Turing complete and it is therefore likely you will end up needing to add the `div` effect in places that you know actually do not diverge. It is nice to know things can not diverge, and nice that in a large majority of the cases you do not have to annotate `div`, but at some point the program becomes complex enough that it is not practical to try to remove `div` from every function. At least this is my opinion. Work hard to make the building blocks non-divergent when you know them to be, but when composing an actual program, accept that you will need to annotate divergence in order to focus on more important things.

As far as your proposed change to where `pretend-decreasing` is annotated. It would have to be something at the level of types I think. Doing a flow analysis to see what values flow to `xs` in the foreach function is an O(n^3) algorithm at minimum, and therefore is impractical to use during type checking.

**Reply by chtenb** [reply]
> As far as your proposed change to where pretend-decreasing is annotated. It would have to be something at the level of types I think. 

That's sounds very reasonable. 

I understand that this part of the feature set is not the most important thing to improve right now, and it wasn't my intention to complain about it. But this came up a few times when dabbling with a few algorithms that can't simply use maps or folds, and I currently solve it by sprinkling `pretend-decreasing` everywhere. It seemed nice if this could be limited to a few core functions, and that other functions understand when a return value is decreasing.

I haven't thought about solving this particular problem with effects, so I'm not sure at this point if that could work. I'll give that some thought. Thanks for the response.

**Reply by TimWhiting** [reply]
Yeah, it's a common complaint I think. One problem with the types approach though is that you have to relate the type of the output to the types of the input (the output is smaller than the input). 
And then there is different meanings of what makes a type `smaller`. For example `next` moves the index one further to the end, which does mean that it makes a vector iterator `smaller`, but only when you are iterating forward. 

For example, if you were implementing a bidirectional iterator based on some result of a user function (bool saying forward or back), then `next` no longer means that the input gets smaller, since you could be going backwards through the vector and `next` actually makes the type `bigger`.

In other words, thinking more about it I don't think the types solution is a viable path forward either. You end up basically needing a fully dependent type system to prove termination in general, and then the type inference isn't guaranteed to terminate unless you restrict the language to not allow non-termination in general, which is a huge restriction! https://en.wikipedia.org/wiki/Dependent_type

**Comment by chtenb** [comment]
To workaround it you can also implement more general recursion utilities, like this

```koka

pub value type tailrec<i,o,r>
  Loop(it : i, loop-arg : o)
  Stop(ret : r)

pub fun vslice/loop( slice : vslice<a>, f : (maybe<(a,vslice<a>)>, o, b) -> e tailrec<vslice<a>,o,r>, loop-arg : o, fixed-arg : b ) : e r
  match f(slice.next, loop-arg, fixed-arg)
    Stop(ret) -> ret
    Loop(it) -> loop(pretend-decreasing(it), f, loop-arg, fixed-arg)
      
fun vslice/foreach( slice : vslice<a>, action : a -> e () ) : e ()
  loop(slice, foreach-rec, (), action)

fun vslice/foreach-rec( it : maybe<(a,vslice<a>)>, _ : (), action : a -> e () ) : e tailrec<vslice<a>,(),()>
  match it
    Nothing -> Stop(())
    Just((x, xs)) -> 
      action(x)
      Loop(xs,())
```

But using `pretend-decreasing` at call sites is so much simpler, and probably more performant too.
